### ISSUE TITLE
Anaconda support

### ISSUE THREAD

**Issue by FlorianWilhelm** [pyscaffold/pyscaffold#135]
Since more and more people use conda packages, a flag `--conda` should create all files necessary to build Anaconda packages.

**Comment by FlorianWilhelm** [comment]
PyScaffold already has an conda package at [conda-forge](https://github.com/conda-forge/pyscaffold-feedstock).

**Comment by FlorianWilhelm** [comment]
Maybe it's best to support this first with the help of an official PyScaffold extension package. I'll drop this as an requirement for core PyScaffold 3.1. Another reasons is that actually no PyScaffold user ever requested this feature nor commented on this ticket.

**Comment by willu47** [comment]
I think this would be very useful. I'm using Pyscaffold for the packages [SALib](/SALib/SALib) and [smif](/nismod/smif), and having trouble with the conda-forge feedstocks for [SALib](/conda-forge/salib-feedstock).

**Comment by FlorianWilhelm** [comment]
I am not sure this issue makes any more sense. The problem is that in order to create a conda package, e.g. for [conda-forge](https://conda-forge.org/), you have an extra repository with a kind of build recipe, aka *feedstock*. [Here is the Python Feedstock as an example](https://github.com/conda-forge/pyscaffold-feedstock). This also means that it makes no sense to create this structure within the current template as it would rather be completely independent in its own repo. Consequently, a `--conda` flag makes no sense. To create a conda feedstock, one could just mention in the docs that everything is explained under https://conda-forge.org/docs/maintainer/adding_pkgs.html.

**Comment by abravalheri** [comment]
Let me see if I got it right:

Anaconda allows users to build and upload packages and this process is described in [conda-build/build-pkgs](https://docs.conda.io/projects/conda-build/en/latest/user-guide/tutorials/build-pkgs.html) and [conda-build/build-pkgs-skeleton](https://docs.conda.io/projects/conda-build/en/latest/user-guide/tutorials/build-pkgs-skeleton.html). These recipes can in theory coexist within the same repository as generated by PyScaffold.

The problem with this approach is that Anaconda is organised in channels, and users can only upload to their own channels, but mixing packages from different channels is generally not a good idea, because there might be clashes and dependency hell...

There are separated communities such as conda-forge (the most prominent one for Python and in special for data science) and bioconda, but these communities have their own procedures for publishing which can be orthogonal with the idea of package code and recipe code coexisting inside the same repository.

Therefore the suggestion is to drop the idea of generating these coexisting recipes since they make little sense and the majority of the people would be interested in conda-forge anyway, is that correct?

If that is the case, I completely agree: we could add this reasoning and point out to conda-forge's documentation...

**Comment by abravalheri** [comment]
Btw @maresb wrote an awesome set of tips for working with conda-forge, so when we document this in the website we most definitely should include a link for his text: https://github.com/pyscaffold/pyscaffold/discussions/422

**Comment by FlorianWilhelm** [comment]
Exactly @abravalheri, this is also my understanding. The `defaults` channel is the one provided by the company Anaconda (formerly known as Continuum Analytics). The main channel for projects beyond `defaults` is `conda-forge` and thus most `environment.yml` will have both defined. Large ecosystems like PyTorch provide their own channel e.g. `pytorch` including their own way of pushing packages to their channel. For me, mixing channels, for instance `defaults`, `conda-forge` and even `pytorch`, always worked as conda is able to resolve this somehow. 

So for OSS maintainers using PyScaffold I think a link to conda-forge's docs and the great explanation by @maresb should be fine as those will use the `conda-forge` channel most likely. 

I don't know if we have PyScaffold users working on proprietary projects which use a private channel and the repo service (or artifact store) provided by Anaconda for money. In my professional life, I have only seen companies using Artifactory, Nexus, etc. which all rely on pip.

**Comment by maresb** [comment]
@abravalheri and @FlorianWilhelm , I don't think I properly articulated the scope of my writeup in #422. Actually I only had in mind that the PyScaffold _developers_ could my writeup in order to generate conda-forge feedstocks for the remaining PyScaffold plugins, e.g. `pyscaffoldext-django`. I didn't mean it as a general intro for PyScaffold _users_.

One could however use my writeup as a basis for conda-forge packaging for PyScaffold users. But then of course one must ask about the use case, and whether it even makes sense, just as @FlorianWilhelm aptly asks.

To summarize what you've already discussed, as @abravalheri points out, one must fist distinguish between the conda-forge channel, the defaults channel, smaller community channels, and private channels, each with their own publishing guidelines.

Personally, I ignore everything except the conda-forge channel. conda-forge offers an open-source packaging ecosystem which, like pip, does Python very well. But it also extends in scope far beyond just Python. In my mind, the advantage of having your Python package on conda-forge is that `conda` becomes aware of your package and can intelligently manage all the dependencies, in particular those dependencies external to Python.

Moreover, conda-forge builds its packages based on official sources, namely PyPI for Python. Thus it is sort of a secondary package publication process, built strictly on top of some other primary package publishing process like PyPI. Since there are already mechanisms for PyScaffold users to publish to PyPI, it makes sense to add a note to the effect of, "if you've already published to PyPI and would further like to publish to conda-forge, look at these references..."

Conclusion: for the conda-forge use case, I don't think it would make sense for PyScaffold to do anything but provide references. Other use cases are, at least for me, uninteresting.

**Comment by maresb** [comment]
What could also be of public interest is the workflow I have developed for a current project of mine. The code is private, so nothing gets published. This is just about how I use conda internally.

I use conda-forge to manage my development environment. The code I write is then delivered to the customer via a Docker image. I have a script which parses my `src/` for imports, works out the corresponding conda/pip packages in my development environment, and writes out the subset of relevant dependencies into an `environment.yaml`. Then in my `Dockerfile` I create a base conda environment and install this `environment.yaml`.

This workflow makes it easy for me to frequently update my development environment. I just need to run my script to update my Docker environment.

I'm leaving out and oversimplifying a bunch of details, but if people are interested I could explain some more.

**Comment by abravalheri** [comment]
Hi @maresb. I got the feeling that would be the case, nevertheless I think the tips you wrote are very useful for anyone trying to publish a project developed with PyScaffold to conda-forge, so I think it is worty mentioning...

What I did for the time being is opening a PR that adds a new chapter in the documentation about conda/Anaconda (#424), trying to summarise what we have been discussing here. I hope that clarifies things.

Regarding the workflow, it is not the first time that we feel like that it would be useful to document the workflows (yours is particularly interesting, automatically parsing imports seems very handy)... @FlorianWilhelm, I wonder if we should include a workflow session in the docs, or at least create a new category in [discussions](https://github.com/pyscaffold/pyscaffold/discussions) so the users can share and discuss their workflows...

**Comment by FlorianWilhelm** [comment]
Thanks for explaining your workflow @maresb. I am not sure I quite understand it. What do you mean that by 
> I use conda-forge to manage my development environment.

So you create a conda environment using the `conda-forge` channel? If you have your conda environment anyway why do you then need to parse your Python `src` folder for imports to automatically create an `environment.yml` from it? Do you have a tool for that as this process seems really to be quite complex. If you import for instance `pytorch` how do you know that it's only available in the `pytorch` channel? 

My workflow, also for customer projects is, I start with `putup --dsproject ...`, edit the generated `environment.yml`, create the conda environment, activate it and develop. Whenever I feel like I need one more library, I install it with conda, add it to `environment.yml` and import it. I admit that there is a repetition in this workflow in the sense that I install a package and also need to add it to the `environment.yml`. Maybe a cleaner way would be to add a new library directly to `environment.yml` and call afterwards:
```
conda env update -f environment.yml --prune
```
which will then install the library, potentially also removing libraries you don't need anymore.

**Comment by FlorianWilhelm** [comment]
Great idea @abravalheri with adding workflows. My suggestion would be to use the Github discussion feature as we would then have living documentation. People can share their workflows, comment on it, etc.

**Comment by abravalheri** [comment]
@FlorianWilhelm I just added a new category [Common Workflows and Practices](https://github.com/pyscaffold/pyscaffold/discussions/categories/common-workflows-and-practices), please feel free to edit the title and description (it is a bit verbose right now 😝).

At some point in time I will add something there.

**Comment by maresb** [comment]
@FlorianWilhelm good questions. 

> What do you mean by "I use conda-forge to manage my development environment."

I develop entirely remotely in the customer's cloud. I chose to build my development environment as a Docker image with [Jupyter Docker Stacks](https://github.com/jupyter/docker-stacks) as a base image. That image uses conda-forge for package management, so I install tons of additional conda-forge packages on top of that. I also define a `pip` section in my `environment.yml` for a few PyPI packages which aren't in conda-forge. (I install [code-server](https://github.com/cdr/code-server), but actually not from conda since that [leads to dependency problems](https://github.com/conda-forge/code-server-feedstock/issues/17).)

> If you have your conda environment anyway why do you then need to parse your Python src folder for imports to automatically create an environment.yml from it?

Otherwise the Docker image I deploy into production for the customer would be huge. I include the productive code with dependencies, and exclude all my exploratory data analysis tools.

> Do you have a tool for that as this process seems really to be quite complex.

I have a hand-written Python script which leverages [`pipreqs`](https://pypi.org/project/pipreqs/) and [`conda_env`](https://github.com/conda/conda/blob/master/conda_env/env.py) to prune to my `src/` requirements. I have to disambiguate the imports as either conda or pip packages. Most of the time it's easy, but I maintain by hand a list of exceptions:
```python
import_to_conda = dict(
    matplotlib="matplotlib-base",
    sklearn="scikit-learn",
    ...
)
```
It sounds really horrible, but in reality it's really not difficult. (Just a few hundred lines very heavily documented and reasonably robust code.)

Unfortunately having a hand-made script locks me in as the administrator. It would be nice to either find a more standard way of doing things or open-source my current method so as to leverage the community.

> If you import for instance pytorch how do you know that it's only available in the pytorch channel?

I typically use TF for gradient descent and PyMC3 for Bayes, so I don't have pytorch installed at the moment. [Docker stacks installs TF via pip](https://github.com/jupyter/docker-stacks/blob/master/tensorflow-notebook/Dockerfile). There is a [painful effort underway](https://github.com/conda-forge/tensorflow-feedstock) to correct the glaring omission of TF2 in conda-forge. But if I were to install pytorch, I'd figure out whatever conda commands were necessary and add them as appropriate to my development Dockerfile.

---

In summary, what I have automated for myself and which (as far as I know) PyScaffold doesn't support is:
1. Dockerization of the project
2. Automatic dependency maintenance

Given my constraint of remote development, these two items are somewhat intertwined and also particular to my case.

I'm sure this could all be done in a simpler way if I were just familiar with the right tools. :stuck_out_tongue:

**Comment by maresb** [comment]
Thanks @abravalheri I copied the discussion over to #426. Let's end the conversation here.

**Comment by abravalheri** [comment]
I think now after @maresb contributions and the doc reorganisation by @FlorianWilhelm we can safely close this issue, right?

**Comment by FlorianWilhelm** [comment]
Yes, thanks @abravalheri.
